packages/vega-parser/src/parsers/guides/axis-config.js

```diff
@@ -1,14 +1,57 @@
 import {Bottom, Top} from './constants';
 import {extend} from 'vega-util';
+import { isSignal } from '../../util';
+import { axisOrientSignalRef, xyAxisSignalRef } from './axis-util';
 
 export default function(spec, scope) {
   var config = scope.config,
       orient = spec.orient,
-      xy = (orient === Top || orient === Bottom) ? config.axisX : config.axisY,
-      or = config['axis' + orient[0].toUpperCase() + orient.slice(1)],
-      band = scope.scaleType(spec.scale) === 'band' && config.axisBand;
+      band = scope.scaleType(spec.scale) === 'band' && config.axisBand,
+      xy,
+      or;
 
-  return (xy || or || band)
+  if (!isSignal(spec.orient)) {
+    xy = (orient === Top || orient === Bottom) ? config.axisX : config.axisY;
+    or = config['axis' + orient[0].toUpperCase() + orient.slice(1)];
+  } else {
+    var axisX = config.axisX || {},
+        axisY = config.axisY || {},
+        axisTop = config.axisTop || {},
+        axisBottom = config.axisBottom || {},
+        axisLeft = config.axisLeft || {},
+        axisRight = config.axisRight || {},
+        axisXYConfigKeys = new Set(
+          Object.keys(config.axisX || {})
+            .concat(Object.keys(config.axisY || {}))
+        ),
+        axisOrientConfigKeys = new Set(
+          Object.keys(config.axisTop || {})
```

Comment:
I now wonder if we can use

```
[
  ...Object.keys(config.axisTop || {}),
  ...Object.keys(config.axisBottom || {}),
  ...Object.keys(config.axisLeft || {}),
  ...Object.keys(config.axisRight || {}),
]
```

(easier to read)

Reply:
Same for the axisXYConfigKeys

Reply:
concat safer for backwards compat? (per discussion below on spread operator)